File: chatuikit/src/main/java/com/cometchat/chatuikit/shared/views/actionbubble/CometChatActionBubble.java
package com.cometchat.chatuikit.shared.views.actionbubble;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.text.SpannableString;
import android.util.AttributeSet;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.Dimension;
import androidx.annotation.StyleRes;

import com.cometchat.chatuikit.CometChatTheme;
import com.cometchat.chatuikit.R;
import com.cometchat.chatuikit.shared.resources.utils.Utils;
import com.cometchat.chatuikit.shared.spans.MentionMovementMethod;
import com.google.android.material.card.MaterialCardView;

/**
 * A custom view that represents a text bubble used for displaying messages in a
 * chat interface. It provides several customization options such as text color,
 * background color, font, border styling, and more. This class extends
 * {@link MaterialCardView} to provide rich material design support.
 */
public class CometChatActionBubble extends MaterialCardView {
    private static final String TAG = CometChatActionBubble.class.getSimpleName();

    private TextView textView;
    private @ColorInt int textColor;
    private @StyleRes int textAppearance;
    private @ColorInt int backgroundColor;
    private int cornerRadius;
    private int strokeWidth;
    private @ColorInt int strokeColor;
    private Drawable backgroundDrawable;
    private @StyleRes int style;

    /**
     * Constructor for initializing the CometChatTextBubble with only a context.
     *
     * @param context The context in which this view is used.
     */
    public CometChatActionBubble(Context context) {
        this(context, null);
    }

    /**
     * Constructor for initializing the CometChatTextBubble with a context and
     * attribute set.
     *
     * @param context The context in which this view is used.
     * @param attrs   The attribute set for styling this view.
     */
    public CometChatActionBubble(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.cometchatTextBubbleStyle);
    }

    /**
     * Constructor for initializing the CometChatTextBubble with a context,
     * attribute set, and a default style.
     *
     * @param context      The context in which this view is used.
     * @param attrs        The attribute set for styling this view.
     * @param defStyleAttr The default style attribute for the view.
     */
    public CometChatActionBubble(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        inflateAndInitializeView(attrs, defStyleAttr);
    }

    /**
     * Initializes the view by inflating the layout and setting up the text bubble.
     *
     * @param attrs        The attribute set for customization.
     * @param defStyleAttr The default style attribute.
     */
    private void inflateAndInitializeView(AttributeSet attrs, int defStyleAttr) {
        Utils.initMaterialCard(this);
        View view = View.inflate(getContext(), R.layout.cometchat_action_bubble, null);
        textView = view.findViewById(R.id.cometchat_action_bubble_text_view);
        textView.setOnLongClickListener(v -> {
            Utils.performAdapterClick(v);
            return true;
        });
        addView(view);
        applyStyleAttributes(attrs, defStyleAttr, 0);
    }

    /**
     * Applies style attributes based on the XML layout or theme.
     *
     * @param attrs        The attribute set containing customization.
     * @param defStyleAttr The default style attribute.
     * @param defStyleRes  The default style resource.
     */
    private void applyStyleAttributes(AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        TypedArray typedArray = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.CometChatActionBubble, defStyleAttr, defStyleRes);
        @StyleRes int style = typedArray.getResourceId(R.styleable.CometChatActionBubble_cometchatActionBubbleStyle, 0);
        typedArray = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.CometChatActionBubble, defStyleAttr, style);
        extractAttributesAndApplyDefaults(typedArray);
    }

    /**
     * Extracts attributes from the given {@link TypedArray} and applies default
     * values.
     *
     * @param typedArray The TypedArray containing the view's attributes.
     */
    private void extractAttributesAndApplyDefaults(TypedArray typedArray) {
        try {
            setTextAppearance(typedArray.getResourceId(R.styleable.CometChatActionBubble_cometchatActionBubbleTextAppearance, 0));
            setTextColor(typedArray.getColor(R.styleable.CometChatActionBubble_cometchatActionBubbleTextColor,
                                             CometChatTheme.getTextColorSecondary(getContext())));
            setCornerRadius(typedArray.getDimensionPixelSize(R.styleable.CometChatActionBubble_cometchatActionBubbleCornerRadius, 0));
            setStrokeWidth(typedArray.getDimensionPixelSize(R.styleable.CometChatActionBubble_cometchatActionBubbleStrokeWidth, 0));
            setStrokeColor(typedArray.getColor(R.styleable.CometChatActionBubble_cometchatActionBubbleStrokeColor,
                                               CometChatTheme.getStrokeColorDefault(getContext())));
            setBackgroundColor(typedArray.getColor(R.styleable.CometChatActionBubble_cometchatActionBubbleBackgroundColor,
                                                   CometChatTheme.getBackgroundColor2(getContext())));
            setBackgroundDrawable(typedArray.getDrawable(R.styleable.CometChatActionBubble_cometchatActionBubbleBackgroundDrawable));
        } finally {
            typedArray.recycle();
        }
    }

    /**
     * Sets the text content of the text bubble.
     *
     * @param text The text to display in the bubble.
     */
    public void setText(String text) {
        textView.setText(text);
    }

    /**
     * Sets the text content of the text bubble using a {@link SpannableString}.
     *
     * @param text The spannable string to display in the bubble.
     */
    public void setText(SpannableString text) {
        textView.setText(text, TextView.BufferType.SPANNABLE);
        textView.setMovementMethod(MentionMovementMethod.getInstance());
    }

    /**
     * Gets the {@link TextView} associated with this text bubble.
     *
     * @return The TextView within the bubble.
     */
    public TextView getTextView() {
        return textView;
    }

    // Getters for testing or direct access if needed
    public int getTextColor() {
        return textColor;
    }

    /**
     * Sets the text color of the text bubble.
     *
     * @param color The color to set for the text.
     */
    public void setTextColor(@ColorInt int color) {
        textColor = color;
        textView.setTextColor(color);
    }    /**
     * Sets the border width of the text bubble.
     *
     * @param width The width of the border, in pixels.
     */
    public void setStrokeWidth(@Dimension int width) {
        strokeWidth = width;
        super.setStrokeWidth(width);
    }

    public int getTextAppearance() {
        return textAppearance;
    }    /**
     * Sets the border color of the text bubble.
     *
     * @param color The color to set for the border.
     */
    public void setStrokeColor(@ColorInt int color) {
        strokeColor = color;
        super.setStrokeColor(color);
    }

    /**
     * Sets the text appearance (style) of the text bubble.
     *
     * @param appearance The resource ID of the text appearance.
     */
    public void setTextAppearance(@StyleRes int appearance) {
        textAppearance = appearance;
        textView.setTextAppearance(appearance);
    }

    public int getBackgroundColor() {
        return backgroundColor;
    }

    /**
     * Sets the background color of the text bubble.
     *
     * @param color The color to set for the background.
     */
    @Override
    public void setBackgroundColor(@ColorInt int color) {
        backgroundColor = color;
        setCardBackgroundColor(color);
    }

    public int getCornerRadius() {
        return cornerRadius;
    }

    /**
     * Sets the corner radius of the text bubble.
     *
     * @param radius The radius to set for the corners, in pixels.
     */
    public void setCornerRadius(@Dimension int radius) {
        cornerRadius = radius;
        setRadius(radius);
    }

    public Drawable getBackgroundDrawable() {
        return backgroundDrawable;
    }

    /**
     * Sets the background drawable for the text bubble.
     *
     * @param backgroundDrawable The drawable to set as the background.
     */
    @Override
    public void setBackgroundDrawable(Drawable backgroundDrawable) {
        this.backgroundDrawable = backgroundDrawable;
        super.setBackgroundDrawable(backgroundDrawable);
    }

    public int getStyle() {
        return style;
    }

    /**
     * Sets the style of the text bubble from a specific style resource.
     *
     * @param style The resource ID of the style to apply.
     */
    public void setStyle(@StyleRes int style) {
        if (style != 0) {
            this.style = style;
            TypedArray typedArray = getContext().getTheme().obtainStyledAttributes(style, R.styleable.CometChatActionBubble);
            extractAttributesAndApplyDefaults(typedArray);
        }
    }

    public int getStrokeWidth() {
        return strokeWidth;
    }

    @Override
    public int getStrokeColor() {
        return strokeColor;
    }




}

File: chatuikit-kotlin/src/main/java/com/cometchat/uikit/kotlin/presentation/shared/messagebubble/actionbubble/CometChatActionBubble.kt
package com.cometchat.uikit.kotlin.presentation.shared.messagebubble.actionbubble

import android.content.Context
import android.text.SpannableString
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.Dimension
import androidx.annotation.StyleRes
import com.cometchat.chat.models.Action
import com.cometchat.uikit.kotlin.R
import com.cometchat.uikit.kotlin.shared.resources.utils.Utils
import com.google.android.material.card.MaterialCardView

/**
 * A custom view that represents an action bubble used for displaying system
 * action messages like "User joined the group" or "User left".
 *
 * This class extends [MaterialCardView] to provide rich material design support.
 *
 * Features:
 * - Centered text display
 * - Support for SpannableString (mentions, formatting)
 * - Distinct visual style to differentiate from regular messages
 * - Customizable styling via XML attributes or programmatically
 */
class CometChatActionBubble @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : MaterialCardView(context, attrs, defStyleAttr) {

    private lateinit var messageTextView: TextView

    // Single style object - NO individual style properties
    private var style: CometChatActionBubbleStyle = CometChatActionBubbleStyle()

    init {
        inflateAndInitializeView(attrs, defStyleAttr)
    }

    /**
     * Initializes the view by inflating the XML layout and binding child views.
     *
     * @param attrs The attribute set for customization.
     * @param defStyleAttr The default style attribute.
     */
    private fun inflateAndInitializeView(attrs: AttributeSet?, defStyleAttr: Int) {
        Utils.initMaterialCard(this)
        LayoutInflater.from(context).inflate(R.layout.cometchat_action_bubble, this, true)
        messageTextView = findViewById(R.id.cometchat_action_bubble_text_view)
        applyStyleAttributes(attrs, defStyleAttr)
    }

    private fun applyStyleAttributes(attrs: AttributeSet?, defStyleAttr: Int) {
        if (attrs == null) return

        var typedArray = context.theme.obtainStyledAttributes(
            attrs, R.styleable.CometChatActionBubble, defStyleAttr, 0
        )
        val styleResId = typedArray.getResourceId(
            R.styleable.CometChatActionBubble_cometchatActionBubbleStyle, 0
        )
        typedArray.recycle()

        typedArray = context.theme.obtainStyledAttributes(
            attrs, R.styleable.CometChatActionBubble, defStyleAttr, styleResId
        )
        // fromTypedArray handles recycling internally
        style = CometChatActionBubbleStyle.fromTypedArray(context, typedArray)
        applyStyle()
    }

    /**
     * Applies all style properties to views.
     */
    private fun applyStyle() {
        // Bubble container styling
        // Action bubbles apply their own background since the outer CometChatMessageBubble is transparent for CENTER-aligned messages
        if (style.backgroundColor != 0) applyBackgroundColor(style.backgroundColor)
        if (style.cornerRadius != 0f) applyCornerRadius(style.cornerRadius)
        if (style.strokeWidth != 0f) applyStrokeWidth(style.strokeWidth)
        if (style.strokeColor != 0) applyStrokeColor(style.strokeColor)
        style.backgroundDrawable?.let { applyBackgroundDrawable(it) }

        // Text styling
        if (style.textColor != 0) applyTextColor(style.textColor)
        if (style.textAppearance != 0) applyTextAppearance(style.textAppearance)
    }

    // ========================================
    // Private Apply Methods
    // ========================================

    private fun applyBackgroundColor(@ColorInt color: Int) {
        setCardBackgroundColor(color)
    }

    private fun applyCornerRadius(radius: Float) {
        setRadius(radius)
    }

    private fun applyStrokeWidth(width: Float) {
        strokeWidth = width.toInt()
    }

    private fun applyStrokeColor(@ColorInt color: Int) {
        strokeColor = color
    }

    private fun applyBackgroundDrawable(drawable: android.graphics.drawable.Drawable) {
        super.setBackgroundDrawable(drawable)
    }

    private fun applyTextColor(@ColorInt color: Int) {
        messageTextView.setTextColor(color)
    }

    private fun applyTextAppearance(@StyleRes appearance: Int) {
        if (appearance != 0) {
            messageTextView.setTextAppearance(appearance)
        }
    }

    // ========================================
    // Public API - Message Setting
    // ========================================

    /**
     * Sets the message to display in the action bubble.
     *
     * @param message The action message
     */
    fun setMessage(message: Action?) {
        if (message == null) {
            setText("")
            return
        }
        setText(message.message ?: "")
    }

    /**
     * Sets the text to display in the action bubble.
     *
     * @param text The text to display
     */
    fun setText(text: String) {
        messageTextView.text = text
    }

    /**
     * Sets the text to display in the action bubble using a SpannableString.
     * Useful for messages with mentions or other formatting.
     *
     * @param text The spannable text to display
     */
    fun setText(text: SpannableString) {
        messageTextView.setText(text, TextView.BufferType.SPANNABLE)
    }

    /**
     * Gets the current text displayed in the action bubble.
     */
    fun getText(): CharSequence = messageTextView.text

    /**
     * Gets the TextView for direct manipulation.
     */
    fun getTextView(): TextView = messageTextView

    // ========================================
    // Public Style Methods
    // ========================================

    /**
     * Sets the style from a style object.
     */
    fun setStyle(style: CometChatActionBubbleStyle) {
        this.style = style
        applyStyle()
    }

    /**
     * Sets the style from a style resource.
     */
    fun setStyle(@StyleRes styleRes: Int) {
        if (styleRes != 0) {
            val typedArray = context.theme.obtainStyledAttributes(
                styleRes, R.styleable.CometChatActionBubble
            )
            // fromTypedArray handles recycling internally
            setStyle(CometChatActionBubbleStyle.fromTypedArray(context, typedArray))
        }
    }

    // ========================================
    // Getters (read from style object)
    // ========================================

    fun getBubbleBackgroundColor(): Int = style.backgroundColor
    fun getBubbleCornerRadius(): Float = style.cornerRadius
    fun getBubbleStrokeWidth(): Float = style.strokeWidth
    fun getBubbleStrokeColor(): Int = style.strokeColor
    fun getTextColor(): Int = style.textColor
    fun getTextAppearance(): Int = style.textAppearance

    // ========================================
    // Setters (update style object + apply)
    // ========================================

    override fun setBackgroundColor(@ColorInt color: Int) {
        style = style.copy(backgroundColor = color)
        applyBackgroundColor(color)
    }

    fun setCornerRadius(@Dimension radius: Int) {
        style = style.copy(cornerRadius = radius.toFloat())
        applyCornerRadius(radius.toFloat())
    }

    fun setBubbleStrokeWidth(@Dimension width: Int) {
        style = style.copy(strokeWidth = width.toFloat())
        applyStrokeWidth(width.toFloat())
    }

    fun setBubbleStrokeColor(@ColorInt color: Int) {
        style = style.copy(strokeColor = color)
        applyStrokeColor(color)
    }

    fun setTextColor(@ColorInt color: Int) {
        style = style.copy(textColor = color)
        applyTextColor(color)
    }

    fun setTextAppearance(@StyleRes appearance: Int) {
        style = style.copy(textAppearance = appearance)
        applyTextAppearance(appearance)
    }

    companion object {
        private val TAG = CometChatActionBubble::class.java.simpleName
    }
}
